src/cmd/apply.rs

```diff
@@ -0,0 +1,152 @@
+use CliResult;
+use CliError;
+use config::{Config, Delimiter};
+use select::SelectColumns;
+use util;
+
+use std::collections::VecDeque;
+use std::io::{Write, BufReader, BufRead};
+use std::process;
+use std::sync::mpsc;
+use std::thread;
+
+static USAGE: &'static str = "
+Runs an external program over specified columns of a csv file.
+
+Usage:
+    xsv apply <column-name> <executable> [options] [<input>]
+
+The program will be started in a subshell, the column will be sent to stdin of the process and we
+will replace the existing column with stdout of the process.
+
+TODO flag to use index instead of name
+
+Common options:
+    -h, --help             Display this message
+    -o, --output <file>    Write output to <file> instead of stdout.
+    -d, --delimiter <arg>  The field delimiter for reading CSV data.
+                           Must be a single character. (default: ,)
+";
+
+#[derive(RustcDecodable)]
+struct Args {
+    arg_column_name: SelectColumns,
+    arg_executable: String,
+    arg_input: Option<String>,
+    flag_output: Option<String>,
+    flag_delimiter: Option<Delimiter>,
+}
+
+pub fn run(argv: &[&str]) -> CliResult<()> {
+    let args: Args = util::get_args(USAGE, argv)?;
+
+    let rconfig = Config::new(&args.arg_input)
+                         .delimiter(args.flag_delimiter)
+                         .select(args.arg_column_name);
+
+    let mut rdr = rconfig.reader()?;
+    let mut wtr = Config::new(&args.flag_output).writer()?;
+
+    let headers = rdr.byte_headers()?;
+    let sel = rconfig.selection(&*headers)?;
+
+    // start up the program
+    let prog = process::Command::new("sh")
+        .arg("-c")
+        .arg(args.arg_executable)
+        .stdout(process::Stdio::piped())
+        .stdin(process::Stdio::piped())
+        .spawn();
+
+    let mut prog = match prog {
+        Ok(p)  => p,
+        Err(e) => return Err(CliError::Io(e))
+    };
+
+    let progout = match prog.stdout.take() {
+        Some(i) => BufReader::new(i),
+        None    => panic!("return a better error here")
```

Comment:
I'm pretty sure you can just `unwrap` (or better, `expect`) here, since you've asked `Command` to capture `stdout`. I believe that should guarantee the presence of `prog.stdout`.